In [1]:
from statsmodels.tsa.arima.model import ARIMA

In [2]:
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf

In [3]:
from statsmodels.tsa.stattools import adfuller

In [4]:
from statsmodels.tsa.seasonal import seasonal_decompose

In [5]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns


sns.set()

#-------------------- 차트 관련 속성 (한글처리, 그리드) -----------
plt.rcParams['font.family']= 'Malgun Gothic'
plt.rcParams['axes.unicode_minus'] = False

#-------------------- 주피터 , 출력결과 넓이 늘리기 ---------------
# from IPython.core.display import display, HTML
from IPython.display import display, HTML

display(HTML("<style>.container{width:100% !important;}</style>"))
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
pd.set_option('max_colwidth', None)

import warnings
warnings.filterwarnings(action='ignore')

In [6]:
# ----------------- 학습 ----------------------
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

# ----------------- 평가 ----------------------
from sklearn.metrics import confusion_matrix

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score, precision_score, recall_score
from sklearn.metrics import precision_recall_curve, plot_precision_recall_curve

from sklearn.metrics import roc_auc_score, roc_curve, plot_roc_curve

# lec14. OLS 회귀분석

* 잔차 제곱 합을 최소화시키는 벡터 선 하나 찾기  
 = 최소 제곱 법

* OLS(Ordinary Least Squares) : 최소제곱(자승)법
* RSS = 잔차제곱합(Y-Y^)^2를  최소화하는 가중치(w = 기울기) 벡터를 구하는 방법

https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html

statsmodels.regression.linear_model.OLS¶
class statsmodels.regression.linear_model.OLS(endog, exog=None, missing='none', hasconst=None, **kwargs)

In [7]:
from statsmodels.regression.linear_model import OLS

In [8]:
# column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
# df = pd.read_csv('./datasets/housing.csv', header=None, delimiter=r"\s+", names=column_names)

# df

In [9]:
from sklearn.datasets import load_boston

df = load_boston()
df

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

In [10]:
col = df['feature_names']
col

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [12]:
# df = pd.DataFrame(df['data'])

# df.columns = col
# df.head()

# OLS 회귀모델

    * y = wX + b
        * w : 회귀계수(coef)
        * b : 편향(bias)

In [20]:
x = [1,3,2,5,6,9,12,23,35,60]
y = [10,21,32,44,56,65,76,89,90,100]

import statsmodels.api as sm
# from statsmodels.regression.linear_model import OLS


x = sm.add_constant(x) #상수항 결합?

model = OLS(y , x)

fit_res = model.fit()
fit_res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.660
Model:                            OLS   Adj. R-squared:                  0.617
Method:                 Least Squares   F-statistic:                     15.52
Date:                Tue, 04 Apr 2023   Prob (F-statistic):            0.00430
Time:                        16:55:42   Log-Likelihood:                -42.623
No. Observations:                  10   AIC:                             89.25
Df Residuals:                       8   BIC:                             89.85
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         37.5081      8.045      4.662      0.002      18.956      56.061
x1             1.3328      0.338      3.939      0.004       0.553       2.113
==============================================================================
Omnibus:                        1.535   Durbin-Watson:                   0.422
Prob(Omnibus):                  0.464   Jarque-Bera (JB):                0.810
Skew:                          -0.256   Prob(JB):                        0.667
Kurtosis:                       1.704   Cond. No.                         31.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## formula 방식

https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.from_formula.html

statsmodels.regression.linear_model.OLS.from_formula¶
classmethod OLS.from_formula(formula, data, subset=None, drop_cols=None, *args, **kwargs)¶

In [24]:

#----------------------------------------------formula 방식

X = [1,3,2,5,6,9,12,23,35,60]
Y = [10,21,32,44,56,65,76,89,90,100]

df = pd.DataFrame({'x' :X,
                   'y':Y})

model = OLS.from_formula('y ~ x ' ,data = df) # 포뮬라 사용법

fit_res = model.fit()
fit_res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.660
Model:                            OLS   Adj. R-squared:                  0.617
Method:                 Least Squares   F-statistic:                     15.52
Date:                Tue, 04 Apr 2023   Prob (F-statistic):            0.00430
Time:                        17:02:50   Log-Likelihood:                -42.623
No. Observations:                  10   AIC:                             89.25
Df Residuals:                       8   BIC:                             89.85
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     37.5081      8.045      4.662      0.002      18.956      56.061
x              1.3328      0.338      3.939      0.004       0.553       2.113
==============================================================================
Omnibus:                        1.535   Durbin-Watson:                   0.422
Prob(Omnibus):                  0.464   Jarque-Bera (JB):                0.810
Skew:                          -0.256   Prob(JB):                        0.667
Kurtosis:                       1.704   Cond. No.                         31.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* y = 1.3328x + 37.5081

최소자승법 y 햇

<pre>

Dep. Variable: 타겟변수명
Model:	OLS
Method:	Least Squares(최소제곱)
No. Observations: 샘플갯수(10건)
Df Residuals: 잔차자유도(샘플갯수-종속변수갯수(y)-독립변수갯수(x))
Df Model: 독립변수 갯수(x)	
Covariance Type: nonrobust(non-constant variance)

R-squared(R2) : 결정계수(모델의 설명력)
   * 전체 데이터에서 회귀모델이 설명할 수 있는 데이터 비율
   * 회귀모델 y = 1.3x + 37은 전체 데이터의 66%를 설명할 수 있다

Adj. R-squared : 보정된 R2
F-statistic: F분포(통계량)
	* F통계량으로 회귀모델 y = 1.3x + 37의 적절성 평가
Prob (F-statistic) : : F분포(통계량) 유의수준

# F통계량 / SSR SSE SST

<pre>
- SSError = SSResidual    : Sum of Squares for Errors (or Residuals) 잔차의 제곱합
- SSReg. SSTr = SSExplain : Sum of Squares for Regression (or Treated) 회귀(예측값) 제곱합
- SSTotal                 : Sum of Squares Total (SSE + SSTr)

- MSE = Mean of Squared Error      = SSError / Df Residuals(8) (총샘플수-독립변수갯수-종속변수갯수)
- MSR = Mean of Sauqred Regression = SSReg   / Df Model(1)     (독립변수의 개수)
- F-statistic (F통계량)            = SSReg.의 평균(MSR) / SSError의 평균(MSE)

## 위의 서머리가 나오게 된 계산

In [35]:
X = [1,3,2,5,6,9,12,23,35,60]
Y = [10,21,32,44,56,65,76,89,90,100]
# Y = 1.3328x + 37.5081

y_mean = np.mean(Y)                         #Y 평균
x_arr = np.array(X)

y_pred = x_arr * 1.3328 + 37.5081           # 회귀식으로 구현한 예측 값
y_true = np.array(Y)                        # 실제값                         

SSError = ((y_true - y_pred)**2).sum()      # 잔차 (실제값 - 예측값) 제곱 총 합

SSReg = ((y_mean - y_pred)**2).sum()        # 편차 ( 평균 - 예측값 ) 제곱 총 합
SST = SSError + SSReg

R2 = SSReg / SST                            # 결정계수 : 모델이 얼마나 잘 설명할 수 있냐

MSE = SSError / 8                           # SSError / Df Residuals(8)
MSR = SSReg / 1                             # SSReg / Df Model(1) (독립변수의 갯수)
F = MSR / MSE                               # SSReg. 의 평균(MSR) / MSE
print(F, R2 , SSError,SSReg,SST)

15.51664628471745 0.6598154386835811 2949.3880276200007 5720.576347620001 8669.96437524


<pre>

Dep. Variable    : 타겟변수명
Model            : OLS
Method           : Least Squares(최소제곱)
No. Observations : 샘플갯수(10건)
Df Residuals     : 잔차자유도(샘플갯수-종속변수갯수(y)-독립변수갯수(x))
Df Model         : 독립변수 갯수(x)
Covariance Type  : nonrobust(non-constant variance)

---------------------------------------------------------

<font color=red><b>R-squared(R2)</b></font>    : 결정계수(회귀식의 설명력) =  SSReg / SST
                   전체 데이터에서 회귀모델이 설명할 수 있는 데이터 비율
                   회귀모델 y = 1.3x + 37은 전체 데이터의 66%를 설명할 수 있다
Adj. R-squared    : 보정된 R2

<font color=red><b>F-statistic</b></font>       : F분포(통계량)
                    F통계량으로 회귀모델 y = 1.3x + 37의 적절성 평가
Prob (F-statistic): F분포(통계량) 유의수준

<font color=red><b>AIC</b></font> BIC           : 손실 가중치 계산 (낮을 수록 좋음)
                  : X피쳐를 이용해 Y를 예측할 수 있는 정도
                    AIC (Akaikie’s Information Criteria)
                    BIC (Bayesian Information Criterion)
                    
---------------------------------------------------------

<font color=red><b>coef</b></font>              : 회귀계수(클수록 y에 영향도가 크다)
std err           : 표준오차 (오차합 / 표준편차) (낮을 수록 좋음)
t                 : t테스트 (평균값이용 : x피쳐가 y에 영향을 주는 정도 : 상관도)
<font color=red><b>P>|t|</b></font>             : 유의수준 (p-value)
[0.025 0.975]     : 신뢰구간

---------------------------------------------------------

Omnibus           : 비대칭도(왜도) 정규성 테스트 값 (크다:정규분포를 따른다)
Prob(Omnibus)     : Omnibus 유의수준

<font color=red><b>Skew</b></font>              : 왜도 (좌우 비대칭도)
<font color=red><b>Kurtosis</b></font>          : 첨도(뾰족)
Durbin-Watson     : (DW검정)잔차의 독립성을 확인할 수 있는 수치
                     1.5 ~ 2.5 사이이면 독립으로 판단(회귀 모형이 적합하다)
                     0 : 잔차들이 양의 자기 상관
                     2 : 독립성(자기 상관이 없다)
                     4 : 잔차들이 음의 자기 상관

* R-squared(R2) : 1이면 좋다
* F-statistic   : 클 수록 분포에 차이가 있다
* AIC           : 낮을 수록 좋음
* coef          : 클수록 y에 영향도가 크다 
* P>|t|         : 0.05보다 낮아야 유의미하다
* Skew          : 정규분포의 왜도 0  
* Kurtosis      : 정규분포의 첨도 0